cache/sampleStore.js

```diff
@@ -0,0 +1,252 @@
+ */
+function cleanSample(s) {
+  let retval = s.get();
+  delete retval.aspect;
+  retval = removeNullsAndStringifyArrays(retval,
+    constants.fieldsToStringify.sample);
+  return retval;
+} // cleanSample
+
+/**
+ * Populate the redis sample store from the db.
+ *
+ * @returns {Promise} which resolves to the list of redis batch responses, or
+ *  false if the feature is not enabled.
+ */
+function populate() {
+  if (!featureToggles.isFeatureEnabled(constants.featureName)) {
+    return Promise.resolve(false);
+  }
+
+  return Sample.findAll()
+  .then((samples) => {
+    const subjectIdx = new Set();
+    const aspectIdx = new Set();
+    const sampleIdx = new Set();
+    const subjectSets = {};
+    const aspectHashes = {};
+    const sampleHashes = {};
+
+    samples.forEach((s) => {
+      const nameParts = s.name.split('|');
+
+      // Generate the redis keys for this aspect, sample and subject.
+      const aspKey = toKey(constants.objectType.aspect,
+        nameParts[1]); // eslint-disable-line no-magic-numbers
+      const samKey = toKey(constants.objectType.sample, s.name);
+      const subKey = toKey(constants.objectType.subject,
+        nameParts[0]); // eslint-disable-line no-magic-numbers
+
+      // Track each of these in the master indexes for each object type.
+      aspectIdx.add(aspKey);
+      sampleIdx.add(samKey);
+      subjectIdx.add(subKey);
+
+      // For creating each individual subject set...
+      if (subjectSets.hasOwnProperty(subKey)) {
+        subjectSets[subKey].push(aspKey);
+      } else {
+        subjectSets[subKey] = [aspKey];
+      }
+
+      // For creating each individual aspect hash...
+      if (!aspectHashes.hasOwnProperty(aspKey)) {
+        aspectHashes[aspKey] = cleanAspect(s.aspect);
+      }
+
+      // For creating each individual sample hash...
+      sampleHashes[samKey] = cleanSample(s);
+    });
+
+    // Batch of commands to create the master indexes..
+    const indexCmds = [
+      ['sadd', constants.indexKey.subject, Array.from(subjectIdx)],
+      ['sadd', constants.indexKey.aspect, Array.from(aspectIdx)],
+      ['sadd', constants.indexKey.sample, Array.from(sampleIdx)],
+    ];
+    const batchPromises = [redisClient.batch(indexCmds).execAsync()];
+
+    // Batch of commands to create each individal aspect hash...
+    const aspectCmds = Object.keys(aspectHashes)
+      .map((key) => ['hmset', key, aspectHashes[key]]);
+    batchPromises.push(redisClient.batch(aspectCmds).execAsync());
+
+    // Batch of commands to create each individal subject set...
+    const subjectCmds = Object.keys(subjectSets)
+      .map((key) => ['sadd', key, subjectSets[key]]);
+    batchPromises.push(redisClient.batch(subjectCmds).execAsync());
+
+    // Batch of commands to create each individal sample hash...
+    const sampleCmds = Object.keys(sampleHashes)
+      .map((key) => ['hmset', key, sampleHashes[key]]);
+    batchPromises.push(redisClient.batch(sampleCmds).execAsync());
```

Comment:
Just to confirm, since we are not using sorted sets, we will have to sort samples in API layer (default by name)?